README.md
# Flutter屏幕适配插件

## 引用
`pubspec.yaml`中添加
```yaml
dependencies:
  flutter_screen_adaptation: ^0.0.2
```
## 用法
- 初始化  
  `await ScreenAdaptation.init(800, 600);` 传入设计稿的宽和高
- 使用  
  `x.wdp`:根据宽度适配；`x.hdp`:根据高度适配；
- 完整用例
```dart
import 'package:flutter/material.dart';

import 'package:flutter_screen_adaptation/flutter_screen_adaptation.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await ScreenAdaptation.init(800, 600);
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
  }


  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Container(width: 400.wdp, height:300.hdp,color: Colors.green,),
        ),
      ),
    );
  }
}

```



CHANGELOG.md
## 0.0.1

* 支持宽&高适配

## 0.0.2

* 修改协议和pubspec.yaml配置

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_screen_adaptation_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


android/src/test/kotlin/io/github/zcys12173/flutter_screen_adaptation/flutter_screen_adaptation/FlutterScreenAdaptationPluginTest.kt
package io.github.zcys12173.flutter_screen_adaptation.flutter_screen_adaptation

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlin.test.Test
import org.mockito.Mockito

/*
 * This demonstrates a simple unit test of the Kotlin portion of this plugin's implementation.
 *
 * Once you have built the plugin's example app, you can run these tests from the command
 * line by running `./gradlew testDebugUnitTest` in the `example/android/` directory, or
 * you can run them directly from IDEs that support JUnit such as Android Studio.
 */

internal class FlutterScreenAdaptationPluginTest {
  @Test
  fun onMethodCall_getPlatformVersion_returnsExpectedValue() {
    val plugin = FlutterScreenAdaptationPlugin()

    val call = MethodCall("getPlatformVersion", null)
    val mockResult: MethodChannel.Result = Mockito.mock(MethodChannel.Result::class.java)
    plugin.onMethodCall(call, mockResult)

    Mockito.verify(mockResult).success("Android " + android.os.Build.VERSION.RELEASE)
  }
}


lib/src/screen_adaptation_impl.dart
import 'package:flutter/services.dart';

class ScreenAdaptation {
  static late double physicalHeight;
  static late double physicalWidth;
  static late double widthScale;
  static late double heightScale;
  static late double _designHeight;
  static late double _designWidth;
  static const _methodChannel = MethodChannel('flutter_screen_adaptation');

  ScreenAdaptation._();

  static Future<Map<String,double>> _getScreenSize() async{
    final map = await _methodChannel.invokeMethod<Map>('getScreenSize');
    return {"height":map?["height"],"width":map?["width"]};
  }

  static Future init(double designWidth,double designHeight) async {
    _designHeight = designHeight;
    _designWidth = designWidth;
    var sizeMap = await _getScreenSize();
    physicalHeight = sizeMap["height"] ?? 0.0;
    physicalWidth = sizeMap["width"] ?? 0.0;
    widthScale =  ScreenAdaptation.physicalWidth/ScreenAdaptation._designWidth;
    heightScale =  ScreenAdaptation.physicalHeight/ScreenAdaptation._designHeight;
  }
}


extension NumberConvert on num{
  double get wdp => this * ScreenAdaptation.widthScale;
  double get hdp => this * ScreenAdaptation.heightScale;
}


ios/Classes/FlutterScreenAdaptationPlugin.swift
import Flutter
import UIKit

public class FlutterScreenAdaptationPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_screen_adaptation", binaryMessenger: registrar.messenger())
    let instance = FlutterScreenAdaptationPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "getScreenSize":
        let screenSize = UIScreen.main.bounds.size
//             let screenScale = UIScreen.main.scale
             let screenWidthInPoints = screenSize.width
             let screenHeightInPoints = screenSize.height
//             print("scall:\(screenScale),width：\(screenWidthInPoints),height:\(screenHeightInPoints)")
             let screenWidthInDp = Float(screenSize.width)
             let screenHeightInDp = Float(screenSize.height)
             result(["width": screenWidthInDp, "height": screenHeightInDp])
    default:
      result(FlutterMethodNotImplemented)
    }
  }
    
    
}


android/src/main/kotlin/io/github/zcys12173/flutter_screen_adaptation/flutter_screen_adaptation/FlutterScreenAdaptationPlugin.kt
package io.github.zcys12173.flutter_screen_adaptation.flutter_screen_adaptation

import android.content.Context
import android.content.res.Resources
import android.util.DisplayMetrics
import android.view.WindowManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlin.math.ceil

/** FlutterScreenAdaptationPlugin */
class FlutterScreenAdaptationPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var windowManager: WindowManager

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_screen_adaptation")
        windowManager =
            flutterPluginBinding.applicationContext.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
        } else if (call.method == "getScreenSize") {
            val metrics = DisplayMetrics()
            windowManager.defaultDisplay.getRealMetrics(metrics)
            val density = Resources.getSystem().displayMetrics.density
            result.success(
                mutableMapOf(
                    Pair("height", ceil(metrics.heightPixels / density)),
                    Pair("width", ceil(metrics.widthPixels / density))
                )
            )
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter_screen_adaptation/flutter_screen_adaptation.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await ScreenAdaptation.init(800, 600);
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
  }


  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Container(width: 400.wdp, height:300.hdp,color: Colors.green,),
        ),
      ),
    );
  }
}
